src/Npgsql.NetTopologySuite/Internal/NetTopologySuiteTypeInfoResolver.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using NetTopologySuite;
+using NetTopologySuite.Geometries;
+using NetTopologySuite.IO;
+using Npgsql.Internal;
+using Npgsql.Internal.Postgres;
+
+namespace Npgsql.NetTopologySuite.Internal;
+
+sealed class NetTopologySuiteTypeInfoResolver : IPgTypeInfoResolver
+{
+    TypeInfoMappingCollection Mappings { get; }
+
+    public NetTopologySuiteTypeInfoResolver(
+        CoordinateSequenceFactory? coordinateSequenceFactory,
+        PrecisionModel? precisionModel,
+        Ordinates handleOrdinates,
+        bool geographyAsDefault)
+    {
+        coordinateSequenceFactory ??= NtsGeometryServices.Instance.DefaultCoordinateSequenceFactory;
+        precisionModel ??= NtsGeometryServices.Instance.DefaultPrecisionModel;
+        handleOrdinates = handleOrdinates == Ordinates.None ? coordinateSequenceFactory.Ordinates : handleOrdinates;
+
+        var reader = new PostGisReader(coordinateSequenceFactory, precisionModel, handleOrdinates);
+        var writer = new PostGisWriter();
+
+        Mappings = new TypeInfoMappingCollection();
+        AddInfos(Mappings, reader, writer, geographyAsDefault);
+        // TODO: Opt-in only
+        AddArrayInfos(Mappings);
+    }
+
+    public PgTypeInfo? GetTypeInfo(Type? type, DataTypeName? dataTypeName, PgSerializerOptions options)
+        => Mappings.Find(type, dataTypeName, options);
+
+    static void AddInfos(TypeInfoMappingCollection mappings, PostGisReader reader, PostGisWriter writer, bool geographyAsDefault)
+    {
+        // geometry
+        mappings.AddType<Geometry>("geometry",
+            (options, mapping, _) => mapping.CreateInfo(options, new NetTopologySuiteConverter<Geometry>(reader, writer)),
+            matchRequirement: !geographyAsDefault ? MatchRequirement.Single : MatchRequirement.DataTypeName);
+
+        mappings.AddType<Point>("geometry",
+            (options, mapping, _) => mapping.CreateInfo(options, new NetTopologySuiteConverter<Point>(reader, writer)),
+            matchRequirement: !geographyAsDefault ? MatchRequirement.All : MatchRequirement.DataTypeName);
```

Comment:
Another option here would be to go for two functions and swap which clr types are added first as a mapping.

As Find will take the first clr type that matches and stores it as a fallback, if there was no `Single` or `DatatypeName` mapping that matched it'll return the fallback.

As a result if all the types are added in priority order, none (except for the default type) need a matchRequirement.